core/store/src/adapter/mod.rs

```diff
@@ -0,0 +1,69 @@
+pub mod flat_store;
+
+use std::io;
+use std::ops::{Deref, DerefMut};
+
+use crate::{Store, StoreUpdate};
+
+enum StoreUpdateHolder<'a> {
+    Reference(&'a mut StoreUpdate),
+    Owned(StoreUpdate),
+}
+
+impl Deref for StoreUpdateHolder<'_> {
+    type Target = StoreUpdate;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            StoreUpdateHolder::Reference(store_update) => store_update,
+            StoreUpdateHolder::Owned(store_update) => store_update,
+        }
+    }
+}
+
+impl DerefMut for StoreUpdateHolder<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self {
+            StoreUpdateHolder::Reference(store_update) => store_update,
+            StoreUpdateHolder::Owned(store_update) => store_update,
+        }
+    }
+}
+
+impl Into<StoreUpdate> for StoreUpdateHolder<'static> {
+    fn into(self) -> StoreUpdate {
+        match self {
+            StoreUpdateHolder::Reference(_) => panic!("converting borrowed store update"),
+            StoreUpdateHolder::Owned(store_update) => store_update,
+        }
+    }
+}
+
+pub trait StoreAdapter {
```

Comment:
I like the fact `store_helpers` is gone 👍 

IMO there's something _off_ with the new API

`StoreAdapter` having a method `flat_store()` and at the same time introducing a`FlatStoreAdapter` implementation of `StoreAdapter` sounds confusing. It feels like the separation is not at the right place.

Should we take a step back and reason about the core idea? 

Do we want to hide "flat storage", "memtries", etc behind an easy to use interface or do we want them to be concrete implementation of a more general "store"? Right now I see an implementation for the former with some sprinkles of the latter.

Reply:
Aah right, that can definitely sound confusing but the idea is to have inter convertibility among all the adapter types. For example if I have an epoch_store_adapter, flat_store_adapter, trie_store_adapter, chain_store_adapter.... I should be able to call xxx_store_adapter.flat_store() to get an instance of flat_store_adapter

Reply:
Flat storage and memtries are to remain as they are right now, this store abstraction is at a much much lower level, right above rocks db.

Reply:
(fyi `flat_store` may sound confusing, I didn't have a better name, but it doesn't refer to `FlatStorage`, but just a wrapper on top of `store`)

Reply:
ahah I see thanks!

on this point do you already know where should flat_store be used vs epoch_store or trie_store..?

Reply:
Umm yeah, there are high level patterns in our codebase (with LOT of exceptions). It should be possible to categorize the store and store update interface needs based on where it's used, like the trie_store, epoch_store etc. etc. that I mentioned. Yes, it does pretty much depend on the location/module where it's used and the set of columns it needs to access.

This is pretty much like a wrapper around Store :)
I wasn't in favor of using traits as with real structs, we have the option to store tiny pieces of data that maybe beneficial, say like a cache.
It may explicitly be useful for our resharding case where we need to map the child shard_uid to parent shard_uid. This can seamlessly be done within the trie_store and trie_store_update structs with just one additional store read! @staffik for context

Reply:
Sorry, to be clear, while `StoreAdapter` is definitely a trait that provides us the inter-convertibility between different types of adapters, the explicit adapters themselves, like `FlatStoreAdapter` are objects (they could have as well been traits)

Reply:
@Trisfald Shreyan and I went through a couple of alternative designs and unfortunately given all the issues with these alternatives it seems the one used in the PR is still relatively the best. I'm a Rust perfectionist and I don't like how this is designed, but practicality is more important I guess.